cluster-autoscaler/core/scaledown/planner/controller.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package planner
+
+import (
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes"
+)
+
+var (
+	errNotFound = errors.New("owner not found")
+)
+
+type replicasInfo struct {
+	targetReplicas, currentReplicas int32
+}
+
+// ControllerReplicasCalculator calculates a number of target and expected replicas for a given controller.
+type ControllerReplicasCalculator interface {
+	GetReplicas(metav1.OwnerReference, *corev1.Pod) (*replicasInfo, error)
+}
+
+type controllerCalculatorImpl struct {
+	listers kubernetes.ListerRegistry
+}
+
+func newControllerCalculator(listers kubernetes.ListerRegistry) ControllerReplicasCalculator {
+	return &controllerCalculatorImpl{listers: listers}
+}
+
+func (c *controllerCalculatorImpl) GetReplicas(ownerRef metav1.OwnerReference, pod *corev1.Pod) (*replicasInfo, error) {
```

Comment:
Does this need to accept a pod? I think `namespace` + `ownerRef` should not only be sufficient, but also less expensive. `GetPodStatefulSets(pod)` and similar functions list all controller objects and iterates through them whereas getting directly (e.g. `c.listers.StatefulSetLister().StatefulSets(namespace).Get(ownerRef.name)`) requires only O(1) lookup.

Reply:
switched to version with ownerRef + namespace. Initially thought we also need to check labels (which GetPodStatefulSets does), but it doesn't seem required